File: user-frontend/src/services/product.ts
import { http } from './http'

export type ProductListItem = {
  id: number
  name: string
  cover_url?: string | null
  price_cent: number
  market_price_cent?: number | null
  merchant_id: number
  merchant_name: string
  sales_count: number
  tags: string[]
}

export type ProductDetail = {
  id: number
  name: string
  description: string
  cover_url?: string | null
  category_id?: number | null
  category_name?: string | null
  status: string
  sales_count: number
  images: string[]
  detail_images: string[]
  merchant: { id: number; name: string }
  skus: Array<{ id: number; name: string; price_cent: number; market_price_cent?: number | null; stock: number }>
  review_summary: { count: number; average_score?: number | null }
}

export type Merchant = {
  id: number
  name: string
  logo_url?: string | null
  announcement?: string | null
}

export type MerchantFollowStatus = {
  merchant_id: number
  followed: boolean
  follower_count: number
}

export type MerchantFollowItem = {
  merchant: Merchant
  followed_at: string
  follower_count: number
}

export type ProductFavoriteStatus = {
  product_id: number
  favorited: boolean
  favorite_count: number
}

export type ProductFavoriteItem = {
  product: ProductListItem
  favorited_at: string
  favorite_count: number
}

export type ProductReview = {
  id: number
  user_id: number
  user_nickname?: string | null
  user_avatar_url?: string | null
  order_id: number
  product_id: number
  score: number
  content: string
  image_urls: string[]
  status: string
}

export type Category = {
  id: number
  name: string
  parent_id?: number | null
  sort_order: number
}

export type PageResult<T> = {
  list: T[]
  page: number
  page_size: number
  total: number
}

export const productService = {
  listProducts(params?: {
    keyword?: string
    category_id?: number
    merchant_id?: number
    min_price_cent?: number
    max_price_cent?: number
    sort_by?: string
    sort_order?: string
    page?: number
    page_size?: number
  }) {
    return http.get<unknown, { data: PageResult<ProductListItem> }>('/products', { params })
  },

  getMerchant(merchantId: number) {
    return http.get<unknown, { data: Merchant }>(`/merchants/${merchantId}`)
  },

  listMerchantProducts(
    merchantId: number,
    params?: {
      min_price_cent?: number
      max_price_cent?: number
      sort_by?: string
      sort_order?: string
      page?: number
      page_size?: number
    },
  ) {
    return http.get<unknown, { data: PageResult<ProductListItem> }>(`/merchants/${merchantId}/products`, { params })
  },

  getMerchantFollowStatus(merchantId: number) {
    return http.get<unknown, { data: MerchantFollowStatus }>(`/merchants/${merchantId}/follow`)
  },

  followMerchant(merchantId: number) {
    return http.post<unknown, { data: MerchantFollowStatus }>(`/merchants/${merchantId}/follow`)
  },

  unfollowMerchant(merchantId: number) {
    return http.delete<unknown, { data: MerchantFollowStatus }>(`/merchants/${merchantId}/follow`)
  },

  listFollowedMerchants(params?: { page?: number; page_size?: number }) {
    return http.get<unknown, { data: PageResult<MerchantFollowItem> }>('/users/followed-merchants', { params })
  },

  getProductFavoriteStatus(productId: number) {
    return http.get<unknown, { data: ProductFavoriteStatus }>(`/products/${productId}/favorite`)
  },

  favoriteProduct(productId: number) {
    return http.post<unknown, { data: ProductFavoriteStatus }>(`/products/${productId}/favorite`)
  },

  unfavoriteProduct(productId: number) {
    return http.delete<unknown, { data: ProductFavoriteStatus }>(`/products/${productId}/favorite`)
  },

  listFavoriteProducts(params?: { page?: number; page_size?: number }) {
    return http.get<unknown, { data: PageResult<ProductFavoriteItem> }>('/users/favorite-products', { params })
  },

  getProduct(productId: number) {
    return http.get<unknown, { data: ProductDetail }>(`/products/${productId}`)
  },

  listProductReviews(productId: number, params?: { page?: number; page_size?: number; score?: number; has_image?: boolean }) {
    return http.get<unknown, { data: PageResult<ProductReview> }>(`/products/${productId}/reviews`, { params })
  },

  listCategories() {
    return http.get<unknown, { data: Category[] }>('/categories')
  },
}

File: app/schemas/product.py
from datetime import datetime

from pydantic import BaseModel, Field


class MerchantCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=80)
    logo_url: str | None = None
    announcement: str | None = None


class MerchantUpdateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=80)
    logo_url: str | None = Field(default=None, max_length=255)
    announcement: str | None = Field(default=None, max_length=255)


class MerchantResponse(BaseModel):
    id: int
    name: str
    logo_url: str | None = None
    announcement: str | None = None

    model_config = {"from_attributes": True}


class MerchantFollowStatusResponse(BaseModel):
    merchant_id: int
    followed: bool
    follower_count: int


class MerchantFollowItemResponse(BaseModel):
    merchant: MerchantResponse
    followed_at: datetime
    follower_count: int


class ProductFavoriteStatusResponse(BaseModel):
    product_id: int
    favorited: bool
    favorite_count: int


class CategoryCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=50)
    parent_id: int | None = None
    sort_order: int = 0


class CategoryUpdateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=50)
    parent_id: int | None = None
    sort_order: int | None = None


class CategoryResponse(BaseModel):
    id: int
    name: str
    parent_id: int | None = None
    sort_order: int

    model_config = {"from_attributes": True}


class SkuCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=80)
    price_cent: int = Field(ge=0)
    market_price_cent: int | None = Field(default=None, ge=0)
    stock: int = Field(default=0, ge=0)
    spec_values: dict = Field(default_factory=dict)


class SkuUpdateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=80)
    price_cent: int | None = Field(default=None, ge=0)
    market_price_cent: int | None = Field(default=None, ge=0)
    stock: int | None = Field(default=None, ge=0)
    spec_values: dict | None = None


class SkuResponse(BaseModel):
    id: int
    name: str
    price_cent: int
    market_price_cent: int | None = None
    stock: int
    spec_values: dict


class ProductCreateRequest(BaseModel):
    merchant_id: int
    category_id: int | None = None
    name: str = Field(min_length=1, max_length=120)
    description: str = ""
    cover_url: str | None = None
    image_urls: list[str] = Field(default_factory=list)
    detail_image_urls: list[str] = Field(default_factory=list)
    skus: list[SkuCreateRequest]


class ProductUpdateRequest(BaseModel):
    category_id: int | None = None
    name: str | None = Field(default=None, min_length=1, max_length=120)
    description: str | None = None
    cover_url: str | None = None
    image_urls: list[str] | None = None
    detail_image_urls: list[str] | None = None


class ProductListItem(BaseModel):
    id: int
    name: str
    cover_url: str | None = None
    price_cent: int
    market_price_cent: int | None = None
    merchant_id: int
    merchant_name: str
    sales_count: int
    tags: list[str] = []


class ProductFavoriteItemResponse(BaseModel):
    product: ProductListItem
    favorited_at: datetime
    favorite_count: int


class ProductDetailResponse(BaseModel):
    id: int
    name: str
    description: str
    cover_url: str | None = None
    category_id: int | None = None
    category_name: str | None = None
    status: str
    sales_count: int = 0
    images: list[str]
    detail_images: list[str] = Field(default_factory=list)
    merchant: MerchantResponse
    skus: list[SkuResponse]
    review_summary: dict = Field(default_factory=lambda: {"count": 0, "average_score": None})


class ProductStatusRequest(BaseModel):
    status: str


class ProductAuditRequest(BaseModel):
    approved: bool


class ProductBatchRequest(BaseModel):
    product_ids: list[int] = Field(min_length=1, max_length=100)


class SkuStockLogResponse(BaseModel):
    id: int
    product_id: int
    sku_id: int
    before_stock: int
    change_quantity: int
    after_stock: int
    change_type: str
    remark: str
    admin_id: int | None = None

    model_config = {"from_attributes": True}
